doc/compiler/JitWriteBarriers.md

```diff
@@ -0,0 +1,339 @@
+         *card = (U_8)CARD_DIRTY;
+      }
+
+      // generational barrier is required if object is old
+      //
+      rememberObject(vmThread, object);
+   }
+```
+
+### j9gc_modron_wrtbar_oldcheck (gencon -Xconcurrentlevel0)
+```
+   /* if value is NULL neither barrier is required */
+   if (NULL != value) {
+      /* Check to see if object is old.
+       * if ((object - vmThread->omrVMThread->heapBaseForBarrierRange0) < vmThread->omrVMThread->heapSizeForBarrierRange0) then old
+       *
+       * Since card dirtying also requires this calculation remember these values
+       */
+      UDATA base = (UDATA)vmThread->omrVMThread->heapBaseForBarrierRange0;
+      UDATA objectDelta = (UDATA)object - base;
+      UDATA size = vmThread->omrVMThread->heapSizeForBarrierRange0;
+      if (objectDelta < size) {
+         /* generational barrier is required if object is old and value is new */
+         /* Check to see if value is new using the same optimization as above */
+         UDATA valueDelta = (UDATA)value - base;
+         if (valueDelta >= size) {
+            /* value is in new space so do generational barrier */
+            rememberObject(vmThread, object);
+         }
+      }
+  }
+```
+
+### j9gc_modron_wrtbar_cardmark_incremental (balanced)
+```
+   /* if value is NULL neither barrier is required */
+   if (NULL != value) {
+      /* Check to see if object is within the barrier range.
+       * if ((object - vmThread->omrVMThread->heapBaseForBarrierRange0) < vmThread->omrVMThread->heapSizeForBarrierRange0)
+       *
+       * Since card dirtying also requires this calculation remember these values
+       */
+      UDATA base = (UDATA)vmThread->omrVMThread->heapBaseForBarrierRange0;
+      UDATA objectDelta = (UDATA)object - base;
+      UDATA size = vmThread->omrVMThread->heapSizeForBarrierRange0;
+      if (objectDelta < size) {
+         /* Object is within the barrier range.  Dirty the card */
+
+         /* calculate the delta with in the card table */
+         UDATA shiftedDelta = objectDelta >> CARD_SIZE_SHIFT;
+
+         /* find the card and dirty it */
+         U_8* card = (U_8*)vmThread->activeCardTableBase + shiftedDelta;
+         *card = (U_8)CARD_DIRTY;
+      }
+   }
+```
+
+### j9gc_modron_wrtbar_cardmark (optavgpause)
+```
+   /* if value is NULL neither barrier is required */
+   if (NULL != value) {
+      /* Check to see if object is old.
+       * if ((object - vmThread->omrVMThread->heapBaseForBarrierRange0) < vmThread->omrVMThread->heapSizeForBarrierRange0) then old
+       *
+       * Since card dirtying also requires this calculation remember these values
+       */
+      UDATA base = (UDATA)vmThread->omrVMThread->heapBaseForBarrierRange0;
+      UDATA objectDelta = (UDATA)object - base;
+      UDATA size = vmThread->omrVMThread->heapSizeForBarrierRange0;
+      if (objectDelta < size) {
+         /* object is old so check for concurrent barrier */
```

Comment:
I've got a question. What is `old` object in non-generational GC policy? What exactly is `objectDelta < size` checking here?